File: Linear_Control_Course/Chapter22/Lesson1/Chapter22_Lesson1.java
import org.apache.commons.math3.complex.Complex;

public class SensitivityPiExample {
    public static void main(String[] args) {
        double Kp = 5.0;
        double Ki = 10.0;
        double J  = 0.01;
        double B  = 0.1;

        Complex j = new Complex(0.0, 1.0);

        for (int k = 0; k <= 40; ++k) {
            double w = Math.pow(10.0, -1.0 + 0.1 * k); // 10^(-1) ... 10^3
            Complex s = j.multiply(w);

            // Plant P(s) = 1 / (J s + B)
            Complex P = new Complex(1.0, 0.0).divide(
                s.multiply(J).add(B)
            );

            // PI controller C(s) = Kp + Ki/s
            Complex C = new Complex(Kp, 0.0).add(
                new Complex(Ki, 0.0).divide(s)
            );

            Complex L = C.multiply(P);
            Complex S = new Complex(1.0, 0.0).divide(
                new Complex(1.0, 0.0).add(L)
            );

            double magS = S.abs();
            double magSdB = 20.0 * Math.log10(magS);

            System.out.printf("w = %.4f, |S(jw)| = %.4f (%.2f dB)%n",
                              w, magS, magSdB);
        }
    }
}

File: Tutorials/Codes/Linear_Control_Course/Chapter22/Lesson1/Chapter22_Lesson1.cpp
#include <iostream>
#include <complex>
#include <vector>
#include <cmath>

int main() {
    using std::complex;
    using std::cout;
    using std::endl;

    double K = 10.0;               // proportional gain
    complex<double> j(0.0, 1.0);   // imaginary unit

    // logarithmic frequency grid
    std::vector<double> w;
    for (int k = 0; k <= 40; ++k) {
        double wk = std::pow(10.0, -1.0 + 0.1 * k); // 10^(-1) ... 10^3
        w.push_back(wk);
    }

    for (double wk : w) {
        complex<double> s = j * wk;
        complex<double> P = 1.0 / (s + 1.0);   // P(s) = 1 / (s + 1)
        complex<double> C = K;                 // C(s) = K
        complex<double> L = C * P;             // loop transfer
        complex<double> S = 1.0 / (1.0 + L);   // sensitivity

        double magS = std::abs(S);
        double magS_dB = 20.0 * std::log10(magS);

        cout << "w = " << wk
             << ", |S(jw)| = " << magS
             << " (" << magS_dB << " dB)" << endl;
    }

    return 0;
}
